### PR TITLE
feat: add sync delete option

### LINKED ISSUE #7124: Allow users to disable deletion for specific objects when application is cascade-deleted
# Summary

Add a new sync option that allows users to disable deletion for specific objects when the application is cascade-deleted.

This will be useful when running patches through Argo CD on objects such as nodes, oauth configs, etc. without this option if the app gets deleted with cascade deletion enabled those objects are deleted.

# Motivation

I need to add a label to a node, I add that label in a yaml file on my repo. Argo CD reads the file and patches the node object adding this new label. Then I delete my Argo CD application with cascade deletion and my node object is deleted. I'd like to have a sync option to disable the deletion for specific objects that I don't want to be deleted on cascade.

# Proposal

Add a new sync-option "Delete=false" that gets checked before deleting the object, if it's found then the object deletion is skipped.

### PR DESCRIPTION
**Pull request by gdsoumya**
Signed-off-by: Soumya Ghosh Dastidar <gdsoumya@gmail.com>

Fixes https://github.com/argoproj/argo-cd/issues/7124 
Depends on https://github.com/argoproj/gitops-engine/pull/507

Thanks to @mvazquezc, the solution proposed works according to my testing. This is the original PR https://github.com/argoproj/argo-cd/pull/7125, but because the PRs were pretty old and had merge conflicts I created a fresh one to close it out.

Note on DCO:

If the DCO action in the integration test fails, one or more of your commits are not signed off. Please click on the *Details* link next to the DCO action for instructions on how to resolve this.

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [x] Does this PR require documentation updates?
* [x] I've updated documentation as required by this PR.
* [ ] Optional. My organization is added to USERS.md.
* [ ] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/tree/master/community#contributing-to-argo)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).

### CODE REVIEW

**Comment by gdsoumya** [comment]
Will need to merge https://github.com/argoproj/gitops-engine/pull/507 and update the go.mod before merging this PR

File: go.mod [hunk 1]
```diff
@@ -256,6 +256,7 @@ require (
 )
 
 replace (
+	github.com/argoproj/gitops-engine => github.com/gdsoumya/gitops-engine v0.0.0-20230213130857-26f153a15602
```
Comment (gdsoumya) [inline go.mod:259]: Will need to merge https://github.com/argoproj/gitops-engine/pull/507 and update the go.mod before merging this PR

Comment (alexmt) [reply go.mod:259]: https://github.com/argoproj/gitops-engine/pull/507 is merged. Please update go.mod

Comment (gdsoumya) [reply go.mod:259]: Thanks! updated.

File: controller/appcontroller.go [hunk 2]
```diff
@@ -941,7 +942,7 @@ func (ctrl *ApplicationController) removeProjectFinalizer(proj *appv1.AppProject
 
 // shouldBeDeleted returns whether a given resource obj should be deleted on cascade delete of application app
 func (ctrl *ApplicationController) shouldBeDeleted(app *appv1.Application, obj *unstructured.Unstructured) bool {
```
Comment (alexmt) [inline controller/appcontroller.go:944]: It should be an easy to test. Can you please add couple unit tests ?

Comment (gdsoumya) [reply controller/appcontroller.go:944]: Added tests

**Review by alexmt — APPROVED** [review]

**Pull request merged**